Explain how to optimise cert-manager for scale

wallrj opened the pull request:
**Preview**: https://deploy-preview-1458--cert-manager-website.netlify.app/docs/devops-tips/scaling-cert-manager/

Fixes: https://github.com/cert-manager/website/issues/551
Fixes: https://github.com/cert-manager/cert-manager/issues/3971
 * https://github.com/cert-manager/website/issues/551
 * https://github.com/cert-manager/cert-manager/issues/3971


@maelvls I've written some general guidance for right sizing cert-manager for both large and small clusters.

maelvls commented on content/docs/devops-tips/large-clusters.md:
````diff
@@ -0,0 +1,170 @@
+which obviates the need for client side throttling.
+Disable the client-side rate limiter as follows:
+
+```yaml
+config:
+  apiVersion: controller.config.cert-manager.io/v1alpha1
+  kind: ControllerConfiguration
+  kubernetesAPIQPS: 10000
+  kubernetesAPIBurst: 10000
+```
````
Where is this YAML supposed to be configured? Is that in the Helm chart?

wallrj commented:
@maelvls @hawksight I've simplified the content and removed all the distracting evidence files.
I've also removed the memory recommendations because I think it is better to let the user simply run `krr` or something on their cluster to learn the best settings.

maelvls commented on content/docs/devops-tips/scaling-cert-manager.md:
```diff
@@ -0,0 +1,134 @@
+Historically this was intended to prevent cert-manager from overwhelming the Kubernetes API server,
+but modern versions of Kubernetes implement [API Priority and Fairness](https://kubernetes.io/docs/concepts/cluster-administration/flow-control/),
+which obviates the need for client side throttling.
+You can increase the threshold of the the client-side rate limiter using the following helm values:
```
```suggestion
You can increase the threshold of the client-side rate limiter using the following helm values:
```

maelvls reviewed (approved):
I enjoyed reading the document. To the point, sticks to the facts, and well sourced, and very well written. Thank you!

It is so good that I'll share this guide on Twitter to talk about cert-manager good practices!

wallrj commented on content/docs/devops-tips/scaling-cert-manager.md:
```diff
@@ -0,0 +1,134 @@
+> 📖 Learn [how to set Certificate defaults automatically](../tutorials/certificate-defaults/README.md), using tools like Kyverno.
+
+
+## Set `revisionHistoryLimit: 1` on all Certificate resources
```
xref:
* https://github.com/cert-manager/cert-manager/issues/3958

inteon commented on content/docs/devops-tips/scaling-cert-manager.md:
```diff
@@ -0,0 +1,134 @@
+>
+> 📖 Learn [how to set `revisionHistoryLimit` when using Annotated Ingress resources](../usage/ingress.md#supported-annotations).
+>
+> 🔗 Read [`cert-manager#3773`: Certificate revision history limit](https://github.com/cert-manager/cert-manager/pull/3773),
```
@wallrj It is quite hard to find a why in that pull request, didn't we discuss in one of our standups that we should set a sane default value?

inteon commented on content/docs/devops-tips/scaling-cert-manager.md:
```diff
@@ -0,0 +1,134 @@
+  apiVersion: controller.config.cert-manager.io/v1alpha1
+  kind: ControllerConfiguration
+  featureGates:
+    AllBeta: true
```
Isn't this the default? Do we have to specifically enable `AllBeta` here?

hawksight replied:
I think so, but I cannot find the exact place in the code that says as such.
It can be good to be explicit in whats enable with helm values, but it's a preference choice.

@wallrj - assuming this is default, I'd drop that flag.. OR just add a comment that this is the default for reference.

wallrj replied:
Agreed. Dropped. I think I assumed it necessary, because I saw it in the E2E setup scripts:

https://github.com/cert-manager/cert-manager/blob/5d0b738b3e35679f1fec3b7a5deac4e7f5a48050/make/e2e-setup.mk#L263-L267

hawksight reviewed (approved):
Easy to read and better docs that what we currently have :)

Left a minor syntax comment.
Also a possible change or clarification on waht @inteon spotted too.

But otherwise I appreciate the links for futher reading / justification. Especially since some of those are to the defaults tutorial 🎉

inteon reviewed (commented):
Thanks @wallrj, great article.
/approve
/lgtm
